wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/JpaWorkspaceActivityDao.java

```diff
@@ -38,71 +40,190 @@
+
+    Consumer<WorkspaceActivity> update;
+    switch (status) {
+      case RUNNING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastRunning(timestamp);
+            };
+        break;
+      case STARTING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStarting(timestamp);
+            };
+        break;
+      case STOPPED:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStopped(timestamp);
+            };
+        break;
+      case STOPPING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStopping(timestamp);
+            };
+        break;
+      default:
+        throw new IllegalStateException("Unhandled workspace status: " + status);
+    }
+
+    withActivity(workspaceId, update);
+  }
+
+  @Override
+  public List<String> findInStatusSince(long timestamp, WorkspaceStatus status)
+      throws ServerException {
+    String queryName = "WorkspaceActivity.get" + firstUpperCase(status.name()) + "Since";
+
+    try {
+      return managerProvider
+          .get()
+          .createNamedQuery(queryName, WorkspaceActivity.class)
+          .setParameter("time", timestamp)
+          .getResultList()
+          .stream()
+          .map(WorkspaceActivity::getWorkspaceId)
+          .collect(Collectors.toList());
+    } catch (RuntimeException e) {
+      throw new ServerException(e.getLocalizedMessage(), e);
+    }
+  }
+
+  @Override
+  public WorkspaceActivity findActivity(String workspaceId) throws ServerException {
+    try {
+      EntityManager em = managerProvider.get();
+      return em.find(WorkspaceActivity.class, workspaceId);
     } catch (RuntimeException x) {
       throw new ServerException(x.getLocalizedMessage(), x);
     }
   }
 
   @Transactional
-  protected List<String> doFindExpired(long timestamp) {
-    return managerProvider
-        .get()
-        .createNamedQuery("WorkspaceExpiration.getExpired", WorkspaceExpiration.class)
-        .setParameter("expiration", timestamp)
-        .getResultList()
-        .stream()
-        .map(WorkspaceExpiration::getWorkspaceId)
-        .collect(Collectors.toList());
+  protected void withActivity(String workspaceId, Consumer<WorkspaceActivity> updater)
+      throws ServerException {
+    withActivity(false, workspaceId, updater);
   }
 
   @Transactional
-  protected void doCreateOrUpdate(WorkspaceExpiration expiration) {
-    final EntityManager manager = managerProvider.get();
-    if (manager.find(WorkspaceExpiration.class, expiration.getWorkspaceId()) == null) {
-      manager.persist(expiration);
-    } else {
-      manager.merge(expiration);
+  protected void withOptionalActivity(String workspaceId, Consumer<WorkspaceActivity> updater)
+      throws ServerException {
+    withActivity(true, workspaceId, updater);
+  }
+
+  private void withActivity(
```

Comment:
`withActivity` looks unclear name for this method. I think methods `doUpdate` and `doUpdateIfExists` would be clearer